### PR TITLE
Switch to go mod

### PR DESCRIPTION
**Pull request by gpaul**
This PR switches argo-cd from using `dep` to using `go modules`.

The following steps were performed, with each step forming a separate commit containing a description and the CLI command that was executed. These steps were performed using go1.13.

Steps:
1. Use `dep` to bring k8s.io dependencies up to date with their existing release branches.
1.1 Update k8s.io/apimachinery to HEAD of the `release-1.14` branch.
1.2 Update k8s.io/api to HEAD of the `release-1.14` branch.
1.3 Update k8s.io/kubernetes to HEAD of the `release-1.14` branch.
1.4 Update k8s.io/client-go to HEAD of the `release-11.0` branch.
2. Switch from `dep` to `go mod` using `go mod init github.com/argoproj/argo-cd`
3. Initialize the `go.lock` file by running `go build ./...`
4. Finally, remove Gopkg.{lock,toml} files.

### CODE REVIEW

**Comment by gpaul** [comment]
This PR needs a bit more work. 

First, I'll wait for https://github.com/argoproj/argo-cd/pull/2326 to land, as it includes a change whereby code generation performed by `make lint` is performed inside the `dev-tools` docker image, using the pinned dependencies specified in the `./hack/Dockerfile.dev` file.

Second, I'll need to remove mention of Gopkg.*, ./vendor, and `dep` from the repo and docs site.

**Comment by alexmt** [comment]
#2326  is in. Looking forward to merge this PR as well!

**Comment by alexmt** [comment]
Spend 4 hours trying to move circle ci and code generation to go modules using your PR 😞. I'm still hoping to get it done. Will send PR to your branch when it is ready.

**Comment by gpaul** [comment]
Oh wow, thanks. Work is intense right now, but I hope to circle back to this PR soon.

**Comment by gpaul** [comment]
Shall I close this in favour of https://github.com/argoproj/argo-cd/pull/2429 ? I started rebasing this PR this morning and saw that you're quite a bit further along, already.

**Comment by alexec** [comment]
I think keep this open. Turns out it pretty hard to move to GO111MODULE and it would be good to keep all the experiments for reference at least.

**Comment by alexec** [comment]
@gpaul what's the chances you can sync with master on this and see if it still builds and passes tests?

**Comment by gpaul** [comment]
Yeah sure, let me try that. It might be a day or two though

**Comment by gpaul** [comment]
I've rebased this branch on `master`, let's see how that fares.

**Comment by alexec** [comment]
To fix CI, in `.circleci/config.yaml` can you try  replace `Gopkg.lock` with `go.sum`?

File: go.mod [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	k8s.io/api v0.0.0-20190816222004-e3a6b8045b0b
+	k8s.io/apiextensions-apiserver v0.0.0-20190404071145-7f7d2b94eca3
+	k8s.io/apimachinery v0.0.0-20190816221834-a9f1d8a9c101
+	k8s.io/client-go v11.0.1-0.20190816222228-6d55c1b1f1ca+incompatible
+	k8s.io/code-generator v0.0.0-20190711102700-42c1e9a4dc7a
+	k8s.io/gengo v0.0.0-20190327210449-e17681d19d3a // indirect
+	k8s.io/klog v0.3.3
+	k8s.io/kube-aggregator v0.0.0-20190711105720-e80910364765
+	k8s.io/kube-openapi v0.0.0-20190502190224-411b2483e503
+	k8s.io/kubernetes v1.15.0-alpha.0.0.20190914015840-8fca2ec50a61
```
Comment (lcostea) [inline go.mod]: I saw that 1.15.7 (which is latest for 1.15) now has support for modules, so we can get rid of the pseudo versions

File: .circleci/config.yml [hunk 2]
```diff
@@ -11,16 +11,16 @@ commands:
             git config --global user.name "Your Name"
             echo "export PATH=/home/circleci/.go_workspace/src/github.com/argoproj/argo-cd/hack:\$PATH" | tee -a $BASH_ENV
             echo "export GIT_ASKPASS=git-ask-pass.sh" | tee -a $BASH_ENV
-  dep_ensure:
+  mod_vendor:
```
Comment (lcostea) [inline .circleci/config.yml]: I am not sure it is a good idea to use mod vendor in ci, it makes it different than the actual release. And it is ignored unless we pass it in `go build` and `go test` with `-mod=vendor`

Comment (gpaul) [reply .circleci/config.yml]: Personally I prefer checking in ./vendor to my repos, but the status quo in argo-cd is to download dependencies in CI (it used to use `dep ensure`, so I switched to using `go mod -vendor`.)

However, IIRC this doesn't quite work, as mod vendor doesn't download the entire dependency, only the transitively depended upon go code in the dependencies. This means that *.proto files, for example, aren't included in the download.

https://github.com/golang/go/issues/26366

This means `go mod -vendor` and `dep ensure` behave differently. The `vend` tool (https://github.com/nomad-software/vend) should be able to help.

**Comment by gpaul** [comment]
This might help: https://github.com/golang/go/issues/26366#issuecomment-467239940

**Comment by lcostea** [comment]
@gpaul I managed to move argo-rollouts to go modules in https://github.com/argoproj/argo-rollouts/pull/331 so I was thinking to give argo-cd a try also. 
I would like to open another PR from my fork, but if you plan to finish work on this one please let me know.

**Comment by gpaul** [comment]
Please do! Shall I go ahead and close this PR?

**Comment by jannfis** [comment]
I have submitted a working migration in #3639 :)

**Comment by gpaul** [comment]
Cool! Closing.

**Pull request closed without merging**